third_party/move/move-compiler-v2/src/env_pipeline/function_checker.rs

```diff
@@ -17,26 +17,54 @@ type QualifiedFunId = QualifiedId<FunId>;
 
 /// check that non-inline function parameters do not have function type.
 pub fn check_for_function_typed_parameters(env: &mut GlobalEnv) {
+    let options = env
+        .get_extension::<Options>()
+        .expect("Options is available");
+    let lambda_params_ok = options.experiment_on(Experiment::LAMBDA_PARAMS);
+    let lambda_return_ok = options.experiment_on(Experiment::LAMBDA_RESULTS);
+    if lambda_params_ok && lambda_return_ok {
+        return;
+    }
+
     for caller_module in env.get_modules() {
         if caller_module.is_primary_target() {
             for caller_func in caller_module.get_functions() {
+                // Check that no functions have function return type
```

Comment:
I've added checks for lambdas in function-type returns and parameters, along with a test case `lambda_returning_lambda.move` and lambda_returning_lambda2.move`, the latter because the error in the the first one hides the remaining error (which can only be seen if the function-typed parameter is unused).